apps/meteor/client/views/navigation/sidepanel/hooks/useRoomMenuActions.ts

```diff
@@ -0,0 +1,142 @@
+	const isNotificationEnabled = !subscription?.disableNotifications;
+
+	const canLeave = ((): boolean => {
+		if (type === 'c' && !canLeaveChannel) {
+			return false;
+		}
+		if (type === 'p' && !canLeavePrivate) {
+			return false;
+		}
+		return !((cl != null && !cl) || ['d', 'l'].includes(type));
+	})();
+
+	const handleHide = useHideRoomAction({ rid, type, name }, { redirect: false });
+	const handleToggleFavorite = useToggleFavoriteAction({ rid, isFavorite });
+	const handleToggleRead = useToggleReadAction({ rid, isUnread, subscription });
+	const handleLeave = useLeaveRoomAction({ rid, type, name, roomOpen });
+	const handleToggleNotification = useToggleNotificationAction({ rid, isNotificationEnabled, roomName: name });
+
+	const isOmnichannelRoom = type === 'l';
+	const prioritiesMenu = useOmnichannelPrioritiesMenu(rid);
+
+	const menuOptions = useMemo(
+		() =>
+			!hideDefaultOptions
+				? [
+						!isOmnichannelRoom && {
+							id: 'hideRoom',
+							icon: 'eye-off',
+							content: t('Hide'),
+							onClick: handleHide,
+						},
+						{
+							id: 'toggleRead',
+							icon: 'flag',
+							content: isUnread ? t('Mark_read') : t('Mark_unread'),
+							onClick: handleToggleRead,
+						},
+						canFavorite && {
+							id: 'toggleFavorite',
+							icon: isFavorite ? 'star-filled' : 'star',
+							content: isFavorite ? t('Unfavorite') : t('Favorite'),
+							onClick: handleToggleFavorite,
+						},
+						canLeave && {
+							id: 'leaveRoom',
+							icon: 'sign-out',
+							content: t('Leave_room'),
+							onClick: handleLeave,
+						},
+					]
+				: [],
+		[
+			hideDefaultOptions,
+			t,
+			handleHide,
+			isUnread,
+			handleToggleRead,
+			canFavorite,
+			isFavorite,
+			handleToggleFavorite,
+			canLeave,
+			handleLeave,
+			isOmnichannelRoom,
+		],
+	);
+
+	const notificationsMenuOptions = useMemo(() => {
+		if (!subscription || hideDefaultOptions || isOmnichannelRoom || !href) {
+			return undefined;
+		}
+
+		return [
+			{
+				id: 'turnOnNotifications',
+				icon: isNotificationEnabled ? 'bell' : 'bell-off',
+				content: isNotificationEnabled ? t('Turn_OFF') : t('Turn_ON'),
+				onClick: handleToggleNotification,
+			},
+			{
+				id: 'notifications',
+				icon: 'customize',
+				content: t('Preferences'),
+				onClick: () => router.navigate(`${href}/push-notifications` as LocationPathname),
+			},
+		];
+	}, [hideDefaultOptions, isNotificationEnabled, subscription, t, router, href, handleToggleNotification, isOmnichannelRoom]);
+
+	if (isOmnichannelRoom && prioritiesMenu.length > 0) {
+		return [
+			{ title: '', items: menuOptions.filter(Boolean) as GenericMenuItemProps[] },
+			{ title: t('Priorities'), items: prioritiesMenu },
+		];
+	}
+
+	return [
+		{ title: '', items: menuOptions.filter(Boolean) as GenericMenuItemProps[] },
+		...(notificationsMenuOptions ? [{ title: t('Notifications'), items: notificationsMenuOptions as GenericMenuItemProps[] }] : []),
+	];
```

Comment:
Suggestion: I noticed the casts to `GenericMenuItemProps[]`, the memos could be improved to return the correct types. Filtering out the falsy values could be done there as well.